crypto/aes/aesccm.c

```diff
@@ -0,0 +1,323 @@
+    while (size > 0 && ctx->pos < AES_BLOCK_SIZE) {
+      ctx->state[ctx->pos] ^= *data;
+      ctx->pos++;
+      data++;
+      size--;
+    }
+
+    if (ctx->pos != AES_BLOCK_SIZE) {
+      return EXIT_SUCCESS;
+    }
+
+    if (aes_encrypt(ctx->state, ctx->state, ctx->encrypt_ctx) != EXIT_SUCCESS) {
+      memzero(ctx, sizeof(*ctx));
+      return EXIT_FAILURE;
+    }
+    ctx->pos = 0;
+  }
+
+  size_t block_count = size >> AES_BLOCK_SIZE_P2;
+  size %= AES_BLOCK_SIZE;
+
+  if (!ALIGN_OFFSET(data, 4)) {
+    while (block_count != 0) {
+      ctx->state32[0] ^= ((uint32_t *)data)[0];
+      ctx->state32[1] ^= ((uint32_t *)data)[1];
+      ctx->state32[2] ^= ((uint32_t *)data)[2];
+      ctx->state32[3] ^= ((uint32_t *)data)[3];
+      if (aes_encrypt(ctx->state, ctx->state, ctx->encrypt_ctx) !=
+          EXIT_SUCCESS) {
+        memzero(ctx, sizeof(*ctx));
+        return EXIT_FAILURE;
+      }
+      data += AES_BLOCK_SIZE;
+      block_count--;
+    }
+  } else {
+    while (block_count != 0) {
+      ctx->state[0] ^= data[0];
+      ctx->state[1] ^= data[1];
+      ctx->state[2] ^= data[2];
+      ctx->state[3] ^= data[3];
+      ctx->state[4] ^= data[4];
+      ctx->state[5] ^= data[5];
+      ctx->state[6] ^= data[6];
+      ctx->state[7] ^= data[7];
+      ctx->state[8] ^= data[8];
+      ctx->state[9] ^= data[9];
+      ctx->state[10] ^= data[10];
+      ctx->state[11] ^= data[11];
+      ctx->state[12] ^= data[12];
+      ctx->state[13] ^= data[13];
+      ctx->state[14] ^= data[14];
+      ctx->state[15] ^= data[15];
+      if (aes_encrypt(ctx->state, ctx->state, ctx->encrypt_ctx) !=
+          EXIT_SUCCESS) {
+        memzero(ctx, sizeof(*ctx));
+        return EXIT_FAILURE;
+      }
+      data += AES_BLOCK_SIZE;
+      block_count--;
+    }
+  }
+
+  while (size > 0) {
+    ctx->state[ctx->pos] ^= *data;
+    ctx->pos++;
+    data++;
+    size--;
+  }
+
+  return EXIT_SUCCESS;
+}
+
+static AES_RETURN cbc_mac_update_zero_padding(cbc_mac_context *ctx) {
+  if (ctx->pos != 0) {
+    if (aes_encrypt(ctx->state, ctx->state, ctx->encrypt_ctx) != EXIT_SUCCESS) {
+      memzero(ctx, sizeof(*ctx));
+      return EXIT_FAILURE;
+    }
+    ctx->pos = 0;
+  }
+  return EXIT_SUCCESS;
+}
+
+static AES_RETURN cbc_mac_final(cbc_mac_context *ctx, uint8_t *mac,
+                                size_t mac_len) {
+  if (ctx->pos != 0 || mac_len > AES_BLOCK_SIZE) {
+    memzero(ctx, sizeof(*ctx));
+    return EXIT_FAILURE;
+  }
+  memcpy(mac, ctx->state, mac_len);
+  memzero(ctx, sizeof(*ctx));
+  return EXIT_SUCCESS;
+}
+
+static AES_RETURN aes_ccm_init(aes_encrypt_ctx *encrypt_ctx,
+                               const uint8_t *nonce, size_t nonce_len,
+                               const uint8_t *adata, size_t adata_len,
+                               size_t plaintext_len, size_t mac_len,
+                               cbc_mac_context *cbc_ctx, uint8_t *ctr_block) {
+  if (mac_len < 4 || mac_len > AES_BLOCK_SIZE || mac_len % 2 != 0) {
```

Comment:
I'm just wondering what is the use case for using a mac with the length of 32 bits. Shouldn't we be more restrictive to avoid misuse? I have no strong opinion, I'm just thinking aloud.

Reply:
This is the range of allowed MAC lengths specified by the NIST standard and they have a test vector for it. The choice of MAC length is a protocol-level matter. I don't see a reason to restrict it here.